weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,383 @@
+        return True
+    else:
+        return False
+
+
+def verify(userid, tfa_response, consume_recovery_code=True):
+    """
+    Verify a 2FA-enabled user's 2FA challenge-response against the stored
+    2FA secret.
+
+    Parameters:
+        userid: The userid to compare the 2FA challenge-response against.
+        tfa_response: User-supplied response string. May be either the Google Authenticator
+        (or other app) supplied code, or a recovery code.
+        consume_recovery_code:  If set to True, the recovery code is consumed if it is
+        valid, otherwise the call to this function is treated as a query to the validity
+        of the recovery code. (Default: Boolean True)
+
+    Returns: Boolean True if 2FA verification is successful, Boolean False otherwise.
+    """
+    # Strip spaces from the tfa_response; they are not valid character, and some authenticator
+    #   implementations display the code as "123 456"
+    tfa_response = tfa_response.replace(' ', '')
+    if len(tfa_response) == LENGTH_TOTP_CODE:
+        # Retrieve the encrypted 2FA secret from `userid`'s login record
+        tfa_secret_encrypted = d.engine.scalar("""
+            SELECT twofa_secret
+            FROM login
+            WHERE userid = %(userid)s
+        """, userid=userid)
+        # Decrypt the 2FA secret to be usable by pyotp
+        tfa_secret = _decrypt_totp_secret(tfa_secret_encrypted)
+        # Validate supplied 2FA response versus calculated current TOTP value.
+        totp = pyotp.TOTP(tfa_secret)
+        # Return the response of the TOTP verification; True/False
+        return totp.verify(tfa_response)
+    elif len(tfa_response) == LENGTH_RECOVERY_CODE:
+        # Check if `tfa_response` is valid recovery code; consume according to `consume_recovery_code`,
+        #  and return True if valid, False otherwise
+        return is_recovery_code_valid(userid, tfa_response, consume_recovery_code)
+    # Other lengths are invalid; return False
+    else:
+        return False
+
+
+def get_number_of_recovery_codes(userid):
+    """
+    Get and return the number of remaining recovery codes for `userid`.
+
+    Parameters:
+        userid: The userid for which to check the count of recovery codes.
+
+    Returns:
+        An integer representing the number of remaining recovery codes.
+    """
+    return d.engine.scalar("""
+        SELECT COUNT(*)
+        FROM twofa_recovery_codes
+        WHERE userid = %(userid)s
+    """, userid=userid)
+
+
+def generate_recovery_codes():
+    """
+    Generate a set of valid recovery codes.
+
+    Character set is defined by `libweasyl.security`, (ASCII+Numbers), limited to uppercase via
+    .upper() for readability.
+
+    Parameters: None
+
+    Returns: A set of length `_TFA_RECOVERY_CODES` where each code is `LENGTH_RECOVERY_CODE`
+    characters in length.
+    """
+    return {security.generate_key(LENGTH_RECOVERY_CODE).upper() for i in range(_TFA_RECOVERY_CODES)}
```

Comment:
This biases the key towards letters, reducing its strength from 119 bits to 102 bits. I’d fix it by changing `security.generate_key` to accept an optional `key_characters` defaulting to `string.ascii_letters + string.digits`.